etl/src/main/java/org/pih/hivmigration/etl/sql/ExamLabResultsMigrator.groovy

```diff
@@ -0,0 +1,81 @@
+package org.pih.hivmigration.etl.sql
+
+class ExamLabResultsMigrator extends ObsMigrator {
+    @Override
+    def void migrate() {
+
+        executeMysql("Create hivmigration_exam_lab_results temp table", '''
+            CREATE TABLE hivmigration_exam_lab_results (
+                source_encounter_id INT,
+                lab_test varchar(80),
+                result varchar(255),
+                test_date date
+            );
+        ''')
+
+        loadFromOracleToMySql('''
+            INSERT INTO hivmigration_exam_lab_results (
+                source_encounter_id, lab_test, result, test_date
+            ) VALUES (?, ?, ?, ?)
+        ''',
+        '''
+            SELECT encounter_id, lab_test, result, test_date 
+            FROM HIV_EXAM_LAB_RESULTS
+        ''')
+
+        create_tmp_obs_table()
+
+        executeMysql("Migrate CD4", '''
+            INSERT INTO tmp_obs (source_encounter_id, concept_uuid, value_numeric)
+            SELECT
+                he.source_encounter_id,
+                concept_uuid_from_mapping('CIEL', '5497'),
+                CASE
+                    WHEN is_number(result) THEN TRIM(result)
+                    WHEN result REGEXP '^\\d*\\s*mm$' THEN extract_number(result)
+                    END
+            FROM hivmigration_exam_lab_results helr
+            JOIN hivmigration_encounters he on helr.source_encounter_id = he.source_encounter_id
+            JOIN encounter e on he.encounter_id = e.encounter_id  -- only migrate lab results corresponding to an encounter that was successfully migrated
+            WHERE
+                lab_test = 'cd4'    
+                AND is_number(result) OR result REGEXP '^\\d*\\s*mm$';  -- e.g. '20 mm'
+        ''')
+
+        executeMysql("Log warning about invalid CD4 values", '''
+            INSERT INTO hivmigration_data_warnings (patient_id, encounter_id, field_name, field_value, note)
+            SELECT
+                hp.person_id,
+                he.encounter_id,
+                'HIV_EXAM_LAB_RESULTS cd4',
+                helr.result,
+                'Invalid value. Not migrated.'
+            FROM hivmigration_exam_lab_results helr
+            JOIN hivmigration_encounters he on helr.source_encounter_id = he.source_encounter_id
+            JOIN hivmigration_patients hp on he.source_patient_id = hp.source_patient_id
+            WHERE lab_test = 'cd4' AND NOT is_number(result) AND NOT result REGEXP '^\\d*\\s*mm$';
```

Comment:
This seems error prone to have to maintain the same query as above, but with the inverse conditional.  It would be better to, say, just look for any rows in the hivmigration_exam_lab_results of type 'cd4' that do not have corresponding rows in tmp_obs, as this represents the actual rows not migrated, and allows us to continuously improve the migration sql without having to ensure we update the corresponding data warning statement.

Reply:
Also, same comment here about "mm" not being valid for CD4

Reply:
+1 to @mseaton 's idea of doing the reverse lookup in tmp_obs.